// @aws-sdk/client-rds-data@3.296.0/dist-es/commands/ExecuteStatementCommand.js
import { getEndpointPlugin } from "@aws-sdk/middleware-endpoint";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ExecuteStatementRequestFilterSensitiveLog, ExecuteStatementResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_restJson1ExecuteStatementCommand, serializeAws_restJson1ExecuteStatementCommand, } from "../protocols/Aws_restJson1";
export class ExecuteStatementCommand extends $Command {
    static getEndpointParameterInstructions() {
        return {
            UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
            Endpoint: { type: "builtInParams", name: "endpoint" },
            Region: { type: "builtInParams", name: "region" },
            UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
        };
    }
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        this.middlewareStack.use(getEndpointPlugin(configuration, ExecuteStatementCommand.getEndpointParameterInstructions()));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "RDSDataClient";
        const commandName = "ExecuteStatementCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: ExecuteStatementRequestFilterSensitiveLog,
            outputFilterSensitiveLog: ExecuteStatementResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_restJson1ExecuteStatementCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_restJson1ExecuteStatementCommand(output, context);
    }
}
