// @aws-sdk/client-iot@3.296.0/dist-es/commands/ListAuditMitigationActionsExecutionsCommand.js
import { getEndpointPlugin } from "@aws-sdk/middleware-endpoint";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListAuditMitigationActionsExecutionsRequestFilterSensitiveLog, ListAuditMitigationActionsExecutionsResponseFilterSensitiveLog, } from "../models/models_1";
import { deserializeAws_restJson1ListAuditMitigationActionsExecutionsCommand, serializeAws_restJson1ListAuditMitigationActionsExecutionsCommand, } from "../protocols/Aws_restJson1";
export class ListAuditMitigationActionsExecutionsCommand extends $Command {
    static getEndpointParameterInstructions() {
        return {
            UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
            Endpoint: { type: "builtInParams", name: "endpoint" },
            Region: { type: "builtInParams", name: "region" },
            UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
        };
    }
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        this.middlewareStack.use(getEndpointPlugin(configuration, ListAuditMitigationActionsExecutionsCommand.getEndpointParameterInstructions()));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "IoTClient";
        const commandName = "ListAuditMitigationActionsExecutionsCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: ListAuditMitigationActionsExecutionsRequestFilterSensitiveLog,
            outputFilterSensitiveLog: ListAuditMitigationActionsExecutionsResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_restJson1ListAuditMitigationActionsExecutionsCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_restJson1ListAuditMitigationActionsExecutionsCommand(output, context);
    }
}
