// @aws-sdk/client-iot@3.296.0/dist-es/commands/ListThingGroupsForThingCommand.js
import { getEndpointPlugin } from "@aws-sdk/middleware-endpoint";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListThingGroupsForThingRequestFilterSensitiveLog, ListThingGroupsForThingResponseFilterSensitiveLog, } from "../models/models_2";
import { deserializeAws_restJson1ListThingGroupsForThingCommand, serializeAws_restJson1ListThingGroupsForThingCommand, } from "../protocols/Aws_restJson1";
export class ListThingGroupsForThingCommand extends $Command {
    static getEndpointParameterInstructions() {
        return {
            UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
            Endpoint: { type: "builtInParams", name: "endpoint" },
            Region: { type: "builtInParams", name: "region" },
            UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
        };
    }
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        this.middlewareStack.use(getEndpointPlugin(configuration, ListThingGroupsForThingCommand.getEndpointParameterInstructions()));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "IoTClient";
        const commandName = "ListThingGroupsForThingCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: ListThingGroupsForThingRequestFilterSensitiveLog,
            outputFilterSensitiveLog: ListThingGroupsForThingResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_restJson1ListThingGroupsForThingCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_restJson1ListThingGroupsForThingCommand(output, context);
    }
}
