// @aws-sdk/client-iot@3.296.0/dist-es/commands/RegisterCertificateWithoutCACommand.js
import { getEndpointPlugin } from "@aws-sdk/middleware-endpoint";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { RegisterCertificateWithoutCARequestFilterSensitiveLog, RegisterCertificateWithoutCAResponseFilterSensitiveLog, } from "../models/models_2";
import { deserializeAws_restJson1RegisterCertificateWithoutCACommand, serializeAws_restJson1RegisterCertificateWithoutCACommand, } from "../protocols/Aws_restJson1";
export class RegisterCertificateWithoutCACommand extends $Command {
    static getEndpointParameterInstructions() {
        return {
            UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
            Endpoint: { type: "builtInParams", name: "endpoint" },
            Region: { type: "builtInParams", name: "region" },
            UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
        };
    }
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        this.middlewareStack.use(getEndpointPlugin(configuration, RegisterCertificateWithoutCACommand.getEndpointParameterInstructions()));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "IoTClient";
        const commandName = "RegisterCertificateWithoutCACommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: RegisterCertificateWithoutCARequestFilterSensitiveLog,
            outputFilterSensitiveLog: RegisterCertificateWithoutCAResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_restJson1RegisterCertificateWithoutCACommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_restJson1RegisterCertificateWithoutCACommand(output, context);
    }
}
